From cb91b1b4707b975f5330c32bfdf306a595ec1a2d Mon Sep 17 00:00:00 2001
From: Gernot Vormayr <gvormayr@gmail.com>
Date: Sat, 13 Sep 2014 19:27:48 +0200
Subject: [PATCH 1/1] gpio: Fix ngpio in gpio-xilinx driver

If one adds gpio-controller; to the chip in the devicetree, then
initialization fails with 'gpiochip_find_base: cannot find free range',
because ngpio is 0. This patch fixes the bug.

Tested on ml507 board.

Signed-off-by: Gernot Vormayr <gvormayr@gmail.com>
---
 drivers/gpio/gpio-xilinx.c | 11 +++++++----
 1 file changed, 7 insertions(+), 4 deletions(-)

diff --git a/drivers/gpio/gpio-xilinx.c b/drivers/gpio/gpio-xilinx.c
index 1248186..4606ad2 100644
--- a/drivers/gpio/gpio-xilinx.c
+++ b/drivers/gpio/gpio-xilinx.c
@@ -197,6 +197,7 @@ static int xgpio_of_probe(struct device_node *np)
 	struct xgpio_instance *chip;
 	int status = 0;
 	const u32 *tree_info;
+	u32 ngpio;
 
 	chip = kzalloc(sizeof(*chip), GFP_KERNEL);
 	if (!chip)
@@ -215,8 +216,9 @@ static int xgpio_of_probe(struct device_node *np)
 	chip->mmchip.gc.ngpio = 32;
 
 	/* Check device node and parent device node for device width */
-	of_property_read_u32(np, "xlnx,gpio-width",
-			      (u32 *)&chip->mmchip.gc.ngpio);
+	status = of_property_read_u32(np, "xlnx,gpio-width", &ngpio);
+	if (status == 0)
+		chip->mmchip.gc.ngpio = (u16)ngpio;
 
 	spin_lock_init(&chip->gpio_lock);
 
@@ -262,8 +264,9 @@ static int xgpio_of_probe(struct device_node *np)
 		chip->mmchip.gc.ngpio = 32;
 
 		/* Check device node and parent device node for device width */
-		of_property_read_u32(np, "xlnx,gpio2-width",
-				     (u32 *)&chip->mmchip.gc.ngpio);
+		status = of_property_read_u32(np, "xlnx,gpio2-width", &ngpio);
+		if (status == 0)
+			chip->mmchip.gc.ngpio = (u16)ngpio;
 
 		spin_lock_init(&chip->gpio_lock);
 
-- 
2.1.0

